### PR TITLE
Rename to SetCredentialRequest::BackupCodeGenerate

### LINKED ISSUE #523: Harmonize SetCredentialRequest Entries
> > * to fit to the other entries in `SetCredentialRequest` enum
> > * https://github.com/kanidm/kanidm/blob/master/kanidm_proto/src/v1.rs#L684
> > * how about renaming `GenerateBackupCode` --> `BackupCodeGenerate`?
> > * could also be a separate ticket, but might also be done here.
> > 
> > What are your thoughts on these?
> 
> Agreed. I wrote that following `GeneratePassword` 😆 (https://github.com/kanidm/kanidm/blob/master/kanidm_proto/src/v1.rs#L673).
> I think we should change both to `BackupCodeGenerate` and `PasswordGenerate`.

I agreed, but this is a seperate issue IMO. Good spot @cuberoot74088

_Originally posted by @Firstyear in https://github.com/kanidm/kanidm/issues/517#issuecomment-873765297_

### PR DESCRIPTION
**Pull request by cuberoot74088**
Fixes #523 

Some renaming as discussed.

### CODE REVIEW

File: kanidm_tools/src/cli/account.rs [hunk 1]
```diff
@@ -23,7 +23,7 @@ impl AccountOpt {
                 AccountCredential::RemoveWebauthn(acs) => acs.copt.debug,
                 AccountCredential::RegisterTotp(acs) => acs.copt.debug,
                 AccountCredential::RemoveTotp(acs) => acs.copt.debug,
-                AccountCredential::GenerateBackupCode(acs) => acs.copt.debug,
+                AccountCredential::BackupCodeGenerate(acs) => acs.copt.debug,
```
Comment (cuberoot74088) [inline kanidm_tools/src/cli/account.rs:26]: Sorry this was not intended. Happy to revert.

File: kanidm_tools/src/cli/account.rs [hunk 2]
```diff
@@ -283,7 +283,7 @@ impl AccountOpt {
                         }
                     }
                 }
-                AccountCredential::GenerateBackupCode(acsopt) => {
+                AccountCredential::BackupCodeGenerate(acsopt) => {
```
Comment (cuberoot74088) [inline kanidm_tools/src/cli/account.rs:286]: Sorry this was not intended. Happy to revert.

File: kanidm_tools/src/opt/kanidm.rs [hunk 3]
```diff
@@ -181,7 +181,7 @@ pub enum AccountCredential {
     GeneratePassword(AccountCredentialSet),
     /// Generate a new set of backup codes.
     #[structopt(name = "generate_backup_codes")]
-    GenerateBackupCode(AccountNamedOpt),
+    BackupCodeGenerate(AccountNamedOpt),
```
Comment (cuberoot74088) [inline kanidm_tools/src/opt/kanidm.rs:184]: Sorry this was not intended. Happy to revert.

**Review by cuberoot74088 — COMMENTED** [review]
Did a few too many replaces. As I said, happy to revert.

File: kanidm_tools/src/cli/account.rs (hunk 1 above)
Comment (Firstyear) [reply kanidm_tools/src/cli/account.rs:26]: No I think consistency is good. So long as it's working I think this is okay :)

**Review by Firstyear — APPROVED** [review]
It does compile/pass tests? do we have a backup code api test in proto_tests? @victorcwai

**Comment by victorcwai** [comment]
> do we have a backup code api test in proto_tests? @victorcwai
No we do not.

File: kanidm_tools/src/cli/account.rs (hunk 1 above)
Comment (victorcwai) [reply kanidm_tools/src/cli/account.rs:26]: We are using different wording for `AccountCredential` and `SetCredentialRequest` though, like:
`AccountCredential::RemoveWebauthn` vs `SetCredentialRequest::WebauthnRemove`

So perhaps we should also rename the other Enum in AccountCredential to follow SetCredentialRequest?

Comment (Firstyear) [reply kanidm_tools/src/cli/account.rs:26]: Yeah, don't see a problem with this :)

**Comment by Firstyear** [comment]
> > do we have a backup code api test in proto_tests? @victorcwai
> 
> No we do not.

Did you want to add these tests @cuberoot74088 ? Or should we open a new issue for it instead?

**Review by victorcwai — APPROVED** [review]

File: kanidm_tools/src/cli/account.rs (hunk 1 above)
Comment (cuberoot74088) [reply kanidm_tools/src/cli/account.rs:26]: So do you want we to do more of these renames?
AccountCredential::RemoveWebauthn vs SetCredentialRequest::WebauthnRemove

How should it be? [Action][CredType] or [CredType][Action]
Have a separate ticket?

**Comment by cuberoot74088** [comment]
I can try to do some t

> > > do we have a backup code api test in proto_tests? @victorcwai
> > 
> > 
> > No we do not.
> 
> Did you want to add these tests @cuberoot74088 ? Or should we open a new issue for it instead?

I don't know yet how to do it, but happy to learn it. 
I think it would be easier to have a separate ticket?

**Comment by Firstyear** [comment]
> I can try to do some t
> 
> > > > do we have a backup code api test in proto_tests? @victorcwai
> > > 
> > > 
> > > No we do not.
> > 
> > 
> > Did you want to add these tests @cuberoot74088 ? Or should we open a new issue for it instead?
> 
> I don't know yet how to do it, but happy to learn it.
> I think it would be easier to have a separate ticket?

How about if you want have a go at the test on this PR, and if it becomes problematic then we make it a new issue?

https://github.com/kanidm/kanidm/blob/master/kanidm_client/tests/proto_v1_test.rs#L750

You can look at this for "inspiration" and by inspiration I mean "largely copy paste". Setup the account with pw + totp + backup codes, and then check:

* Good backup code
* An incorrect backup code

You do NOT need to test that the backup code is removed after you use it because this is async and may have a race condition in the tests, and it's already tested in https://github.com/kanidm/kanidm/blob/master/kanidmd/src/lib/idm/authsession.rs#L2107

**Comment by cuberoot74088** [comment]
@Firstyear Thanks for the hints, very useful! I created the test as proposed using a good and a bad backup code in combination with totp. Is there a need to have this also for webauthn? I don't think so. 

Anything else? If you are happy, feel free to merge it.

**Comment by Firstyear** [comment]
@cuberoot74088 This looks awesome! Great work! 

I'd like @victorcwai to review as well, but I see nothing at the moment that will block merging :)

**Review by victorcwai — APPROVED** [review]
Approved as well

**Comment by Firstyear** [comment]
@cuberoot74088 Again, thank you very much!

**Pull request merged**